api/client/src/main/java/org/projectnessie/client/http/impl/BaseHttpRequest.java

```diff
@@ -27,65 +27,170 @@
+  protected ResponseContext processResponse(
+      URI uri, Method method, Object body, RequestContext requestContext) {
+    try {
+      ResponseContext responseContext = sendAndReceive(uri, method, body, requestContext);
+      processResponseFilters(responseContext);
+      mimicUrlConnectionBehavior(method, responseContext, uri);
+      return responseContext;
+    } catch (ProtocolException e) {
+      throw new HttpClientException(
+          String.format("Cannot perform request against '%s'. Invalid protocol %s", uri, method),
+          e);
+    } catch (MalformedURLException e) {
+      throw new HttpClientException(
+          String.format("Cannot perform %s request. Malformed Url for %s", method, uri), e);
+    } catch (SocketTimeoutException e) {
+      throw new HttpClientReadTimeoutException(
+          String.format(
+              "Cannot finish %s request against '%s'. Timeout while waiting for response with a timeout of %ds",
+              method, uri, config.getReadTimeoutMillis() / 1000),
+          e);
+    } catch (IOException e) {
+      throw new HttpClientException(
+          String.format("Failed to execute %s request against '%s'.", method, uri), e);
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    }
   }
 
-  protected void cleanUp() {
+  protected abstract ResponseContext sendAndReceive(
+      URI uri, Method method, Object body, RequestContext requestContext)
+      throws IOException, InterruptedException;
+
+  protected void processRequestFilters(RequestContext requestContext) {
+    if (!bypassFilters) {
+      config.getRequestFilters().forEach(requestFilter -> requestFilter.filter(requestContext));
+    }
+  }
+
+  protected void processResponseFilters(ResponseContext responseContext) {
+    if (!bypassFilters) {
+      config.getResponseFilters().forEach(responseFilter -> responseFilter.filter(responseContext));
+    }
+  }
+
+  /**
+   * This mimics the (weird) behavior of java.net.HttpURLConnection.getResponseCode() that throws an
+   * IOException for some status codes.
+   */
+  protected void mimicUrlConnectionBehavior(
+      Method method, ResponseContext responseContext, URI uri) {
+    Status status = responseContext.getStatus();
+    if (status.getCode() >= 400) {
+      throw new HttpClientResponseException(
+          String.format("%s request to %s failed with HTTP/%d", method, uri, status.getCode()),
+          status);
+    }
+  }
+
+  protected void processCallbacks(
+      RequestContext requestContext,
+      @Nullable ResponseContext responseContext,
+      @Nullable RuntimeException error) {
+    List<BiConsumer<ResponseContext, Exception>> callbacks = requestContext.getResponseCallbacks();
+    if (callbacks != null) {
+      for (BiConsumer<ResponseContext, Exception> callback : callbacks) {
+        try {
+          callback.accept(responseContext, error);
+        } catch (Exception e) {
+          LOGGER.warn("Callback invocation failed", e);
```

Comment:
Would it make sense to collect all callback exceptions and re-throw as "suppressed" under a new `RuntimeException`? As far as I can see, callback exceptions used to bubble up to the client caller before this PR.